Update desgin document with known issues section (#327)
exasol/python-toolbox: doc/design.rst

      - Description
    * - python-environment
      - Sets up an appropriate poetry based python environment

Known Issues
------------

This section documents flaws, sins, and known issues with the current design and/or its current implementation that were either known upfront or surfaced through the course of implementing it. Additionally, it attempts to explain why certain choices were made at the time, so one can better understand whether it may be reasonable to make changes now or in the future.

Passing files as individual arguments on the CLI
++++++++++++++++++++++++++++++++++++++++++++++++

**Description:**

> As of today selection of python files for litting formatting etc. is done by passing all relevant python files as individual argument(s)
> to the tools used/invoked by the python toolbox.

Reviewer: ```suggestion
As of today selection of python files for linting, formatting etc. is done by passing all relevant python files as individual argument(s)
to the tools used/invoked by the python toolbox.
```

**Downsides:**

- Most shells have limitations on the number of arguments and their length.

> - Noisey output, making it hard to decipher the actual command.

Reviewer: ```suggestion
- Noisy output, making it hard to decipher the actual command.
```
or "verbose"?

- Not ideal for all use cases.

**Rationale/History:**

- The current method of passing files as individual arguments by default offers ease in collection and filtering. It also allows users to simply provide or replace the selection mechanism fairly easily.

> - Every tool used by the toolbox (e.g., `black`, `isort`) used to support passing files by argument. However, not all of them provided the same mechanism for selection or deselection patterns (e.g., "glob").

Reviewer: ```suggestion
- Every tool used by the toolbox (e.g., `black`, `isort`) used to support passing files by argument. However, not all of them provided the same mechanism for selection or deselection patterns (e.g. "glob").
```

**Ideas/Solutions:**

- Develop a wrapper that allows for different selection mechanisms

Inconsistent Naming
+++++++++++++++++++

**Description:**

The naming is not consistent across the project name (python-toolbox) and the PyPI package name (exasol-toolbox).

**Downsides:**

- Misalignment between the PyPI package name and the project name causes confusion when discussing or referring to the project/package.

**Rationale/History:**

- Initially, this was a proof of concept (POC) to verify a few ideas, and the naming was not well thought out at the time.
- Later, when publishing the first package for distribution, the project name was unavailable on PyPI, resulting in a different name being used on PyPI.

**Ideas/Solutions:**

> - `Issue-325 <https://github.com/exasol/python-toolbox/issues/325>`_

Reviewer: ```suggestion
- Consistently rename project to ``exasol-python-toolbox``: `Issue-325 <https://github.com/exasol/python-toolbox/issues/325>`_
```

- Configurations overlap or conflict with unclear priorities.
- Tool leakage (e.g., the ``[isort]`` section in ``pyproject.toml``).
  (If everything were done via toolbox config file(s), backing tools could be swapped more easily).

**Rationale/History:**

- Initial decisions aimed to simplify individual adjustments in the projects until we had a better understanding of what needed to be configured.
- Scattering configuration across various files and tools was a hasty decision to expedite development and accommodate various tools.

**Ideas/Solutions:**

Currently used methods to configure toolbox-based projects:

#. Project configuration: ``noxconfig.py``
#. Tool-specific configuration files or sections in ``pyproject.toml``
#. Implementing plugin extension points
#. Overwriting nox tasks with custom implementations
#. Replacing with customized workflows of the same name (only applicable for action/workflows)

Refinement:

- Centralize all toolbox based configurations in a toolbox config file (``noxconfig.py``).
- Rename the toolbox config file from ``noxconfig.py`` to a more appropriate name that reflects its purpose.
- Document configuration hierarchy and usage.

> 

Author: What is an extension workflow from your POV, could you please elaborate?

> 

Reviewer 2: an empty workflow which can be overwritten by the project and which is called from our default workflows

> 

Author: I can see that. I don't think these issues should be mixed with general configuration, though. Personally, I would distinguish between project configuration and setup versus workflow/gh-action setup.

> 

Reviewer 2: ok, that is fair

Nox Task Runner
+++++++++++++++

**Description:**
While Nox isn't a perfect fit, it still meets most of our requirements for a task runner.

**Downsides:**

- Imports over top-level modules are problematic as all contained tasks are imported.
- Passing and receiving additional arguments to a task is clunky.
- The default behavior of creating a venv for tasks is undesirable.
- Nox does not support grouping.

**Rationale/History:**

Why Nox was choosen:

> - No Additional Language(s) Required: There was no need to introduce extra programming languages or binaries, simplifying the development process.

Reviewer: ```suggestion
- No additional language(s) required: There was no need to introduce extra programming languages or binaries, simplifying the development process.
```

- Python-based: Being Python-based, Nox can be extended and understood by Python developers.
- Python code: As Nox tasks are defined via Python code, existing scripts can be reused and code can be shared easily.
- Simplicity: Nox is relatively "small" in functionality, making it somewhat simple to use and understand.

**Ideas/Solutions:**

Grouping:

Since Nox doesn't natively support task grouping, we need a strategy to group commands.
Therefore, a naming convention to indicate grouping should be adopted.

    Suggestion: Groups will be separated using a :code:`:` (colon) because :code:`-` (dash) might already be used within task names.

Imports:

Consider modularizing tasks to handle top-level imports better.

Others Issues:

> Generally, one may consider addressing the other issues by choosing another task runner or creating a small set of CLI tools and extension points manually provided by the toolbox.

Reviewer 2: Oh, so you mean a wrapper which calls the nox sessions and handles the argument passing more elegant or from ground up a new cli tool. In a new cli tool we would need to think about adding in a project new commands.

Author: Both:

1. Create a wrapper or wrapper tool to uniformly collect relevant files.
2. For example, create a set of CLI tools to replace or wrap the nox tasks, providing a more convenient execution interface.

Poetry for Project Management
+++++++++++++++++++++++++++++

While poetry was and is a good choice for exasol project, dependency, build tool etc. "most recently"
`uv <https://docs.astral.sh/uv/>`_ has surffeced and made big advanced. Looking at uv it addresses additional itches with
our projects and therefore in the long run it may be a good idea to migrate our project setups to it.

> Use poetry for project, build and depenency management.

Reviewer 2: This sentence doesn't make sense

> While poetry was and is a good choice for exasol project, dependency, build tool etc. "most recently"
> `uv <https://docs.astral.sh/uv/>`_ has surffeced and made big advanced. Looking at uv it addresses additional itches with
> our projects and therefore in the long run it may be a good idea to migrate our project setups to it.
> Use poetry for project, build and depenency management.

Reviewer: ```suggestion
While poetry was and is a good choice for Exasol project, dependency, build tool etc. "most recently"
`uv <https://docs.astral.sh/uv/>`_ has surfaced and made big advanced. Looking at uv it addresses additional itches with
our projects and therefore in the long run it may be a good idea to migrate our project setups to it.
Use poetry for project, build and dependency management.
```

**Downsides:**

- Two tools and an aligned configuration of them are required to cleanly and correctly format the codebase.
- Code needs to be processed at least twice as we apply two individual tools.
- The performance of Black and Isort is okay but not great compared to other tools.

**Rationale/History:**

- Black and Isort have been used because they are battle-tested and widely used
- When we opted for Black and Isort, ``ruff`` wasn't "a thing" yet and at best in its early stages.
- Black and Isort already have been known by most python devs when we where selecting the tools

**Ideas/Solutions:**

As `Ruff <https://docs.astral.sh/ruff/>`_ is fairly stable and also tested and used by many Python projects
we should consider transitioning to it.

Advantages:

- Well-tested
- Widely used
- Excellent performance
- Single tool for imports and formatting the codebase
- Simplifies adopting ruff for linting

Pylint
++++++

**Description:**
We are currently using Pylint instead of Ruff.

**Downsides:**

- Pylint is slower and less usable in pre-commit hooks

> - It is an additional tool, therefore at leas one more processing run of the code is required

Reviewer 2: ```suggestion
- It is an additional tool, therefore at least one more processing run of the code is required
```

> - It is an additional tool, therefore at leas one more processing run of the code is required

Reviewer: ```suggestion
- It is an additional tool, therefore at least one more processing run of the code is required
```

> - No LSP (e.g. compare to `ruff lsp`)

Reviewer 2: Interesting

> - No LSP (e.g. compare to `ruff lsp`)

Reviewer: ```suggestion
- No support for Language Server Protocol (LSP, e.g. compare to `ruff lsp`)
```

**Rationale/History:**

- Well known
- Pylint provides built-in project score/rating
- Project score is good for improving legacy code bases which haven't been linted previously
- Plugin support

**Ideas/Possible Solutions:**

Replacing Pylint with Ruff for linting would provide significant performance improvement. Additionally, Ruff offers an LSP and IDE integrations and is widely used these days. Additionaly there would be an additional synergy if we adopt ruff for formatting the code base.

Transitioning to Ruff requires us to adjust the migration and improvement strategies for our projects:

- Currently, our codebase improvements are guided by scores. However, with Ruff, a new approach is necessary. For example, we could incrementally introduce specific linting rules, fix the related issues, and then enforce these rules.

> - The project rating and scoring system will also need modification. One possiblity would be to run Ruff and Pylint in parallel, utilizing Pylint solely for rating and issue resolution while Ruff is incorporated for linting tasks.

Reviewer 2: Scoring is fortunately not need in precommit , only in CI. Are ruff and pylint compatible in their rules?

Author: AFIK, ruff has configuration compatibility with `flake8` and various `flake8` extensions, `isort`, and `pyupgrade`. There is also compatibility with various `pep8` rules. Therefore, there is definitely a significant overlap with `pylint` but I am not sure if it is fully compatible.

> - The project rating and scoring system will also need modification. One possiblity would be to run Ruff and Pylint in parallel, utilizing Pylint solely for rating and issue resolution while Ruff is incorporated for linting tasks.

Reviewer: Could Sonar be an alternative for open-source projects? Sonar supports incremental linting, i.e. findings in unmodified existing code are not considered as build breaker.
Sonar also provides code coverage and PR merge gates via branch protection rules.

Author: I think Sonar could be used or integrated, but solely depending on a web service for this is not a good choice, in my opinion. That said, Sonar supports input from third-party tools like `ruff` and `pylint`. See [here](https://docs.sonarsource.com/sonarqube-cloud/enriching/external-analyzer-reports/#python).

Security Linter
+++++++++++++++

**Description:**

> As of today, the security linter does not fail if it has findings. This was intentionally done to simplify integration and adoption of the tool. Developers can still use the results to improve and find issues within the codebase, and additionally, a rating will be generated to provide some guidance on which projects need attention.

Reviewer 2: This reminds me on the talk about defect dojo. https://media.ccc.de/v/38c3-vulnerability-management-with-defectdojo
I think, a future feature for the security linter could be that it creates a tickets if found on the main branch, but maybe this is actually not needed if we run it always and make it fail at some point.

Author: Good point but I think there is quite a bit of stuff to do before it generally can be enforced.

Reviewer 2: agree

**Downsides:**
- No enforced safeguard on introducing potential security issues

**Rationale/History:**
- Simplify adoption into projects
- First step to introduce tooling and make the current state/rating visible

**Ideas/Possible Solutions:**

> Define a strategy to address potential security issues in projects. Once this has been done, enforce the immediate addressing of potential security issues in the codebase upon introduction.

Reviewer 2: Do we have a config option which we can set per project to enable the enforcement

Author: could be introduced if the following [line](https://github.com/exasol/python-toolbox/blob/30274eb48a587403dedeabe4d83b7bc50754ad8d/exasol/toolbox/nox/_lint.py#L62) will be made configurable.

> Define a strategy to address potential security issues in projects. Once this has been done, enforce the immediate addressing of potential security issues in the codebase upon introduction.

Reviewer: ```suggestion
* Define a strategy to address potential security issues in projects. Once this has been done, enforce the immediate addressing of potential security issues in the codebase upon introduction.
* Allow excluding individual findings in projects until they are fixed.
```

Workflows Dependency Structure
++++++++++++++++++++++++++++++

**Description:**
Undocumented workflow interdependencies and structure

**Downsides:**
- Hard to customize if one does not understand the overall setup and dependencies

**Rationale/History:**
- Simplify development during the discovery phase (what is needed, how to implement, adjust to discovered needs)
- Ideally, all workflows will be integrated and use a standard setup (part of the customization can also be done in the called nox tasks)

**Ideas/Possible Solutions:**

> 

Reviewer 2: See https://github.com/exasol/python-toolbox/pull/327/files#r1901815074

Reviewer 2: Do we have a ticket with our sketched future workflow structure
